typescript-for-haskellers/article.md

```diff
@@ -0,0 +1,916 @@
+And when you will try to pass value of another tagged type, compiler will say that two unique symbols are different.
+
+```typescript
+function sendMessage(email: Email);
+
+sendMessage("message"); // Argument of type 'string' is not assignable to parameter of type 'Email'.
+sendMessage("St. Petersburg" as City); // Argument of type 'City' is not assignable to parameter of type 'Email'.
+
+sendMessage("email@gmail.com" as Email); // Ok
+```
+
+Anyway, this is not an ideal solution for newtypes.
+There is a [comment](https://github.com/Microsoft/TypeScript/issues/4895#issuecomment-401067935) in issue with a better solution.
+And there is also a [library](https://github.com/sveseme/taghiro) which provide more pretty implementation for them.
+
+### Unit types, Unions, Data types, and Pattern Matching
+
+Haskell ADT is a constantly used functionality of language.
+It allows you to build your own big types from small blocks.
+And with pattern matching it is easy to get access to this data.
+
+Unfortunately there is no same way to build ADT, but we will start from the easiest enums and move to the more complex things showing what we can do using existing type system in TypeScript.
+
+#### Unit types
+
+Unit types are subtypes of primitive types that contain exactly one primitive value.
+You can use it like enums setting unique string values.
+Or even use different values types.
+
+```typescript
+type Result = true | "error" | 5;
+```
+
+You can pass it to the function and match by a concret type.
+
+```typescript
+const resultInterpreter = (result: Result): string => {
+  if (result === true) {
+    return "true";
+  }
+  if (result === "error") {
+    return result;
+  }
+  return "five";
+};
+```
+
+#### Unions
+
+Moving forward the next feature is Unions.
+Unions are not tagged, so it is just a condition of possible types.
+
+```typescript
+type Result = string | number | (() => string);
+```
+
+Here and in the example with Unit types if you already matched by some type compiler will know that it will not be present in the future.
+As in the first `if` we matched `typeof result === "number"` then we know that in the code below it may be only `function` or `string`.
+And we can simply run `lenght.toString()` on this value as both `function` and `string` have property `lenght` even if `number` doesn't have such property.
+But we can't `call` this value as `string` is not callable and we can do it only after matching with `string`.
+
+```typescript
+const resultInterpreter = (result: Result): string | undefined => {
+  if (typeof result === "number") {
+      return result.toFixed();
+  }
+
+  // Okay: return result.lenght.toString(); 
+  // Error: return result();
+
+  if (typeof result === "string") {
+      return result.toLocaleUpperCase();
+  }
+
+  return result();
+};
+```
+
+#### Discriminated Unions or Datatype analogue
+
+Going deeper using Unions, it is possible to construct data type analog.
```

Comment:
By going deeper with unions, it is possible to construct an analogue of Haskell data types?